posts/2022/02/making-gpus-accessible-to-pydata-ecosystem-via-array-api.md

```diff
@@ -0,0 +1,486 @@
+arrays. This is not possible with the Array API's dispatching via `get_namespace`. For this demo we have
+added a workaround, until we have implemented the support for `uarray` based backed in SciPy. The
+workaround looks something like this for cupy:
+
+```python
+xp, _ = get_namespace(<input_arrays>)
+
+if xp.__name__ == 'cupy.array_api':
+  from cupyx.scipy import sparse as cupy_sparse
+  coo_matrix =  cupy_sparse.coo_matrix
+else:
+  coo_matrix = scipy.sparse.coo_matrix
+
+graph = coo_matrix(<input_arrays>)
+```
+
+Note that, this workaround is only valid for NumPy/CuPy arrays.
+
+
+#### 2. Not implemented functions
+
+There are still a bunch of functions in the NumPy API which don't
+have a direct equivalent in the Array API, either due to the fact that those functions are not
+generic enough to be defined in the Array API or we haven't got around implementing those.
+
+Some examples of unimplemented functions that we encountered during this excercise:
+
+
+* `np.atleast_3d`
+* `np.strides`
+* `np.real`
+* `np.clip`
+
+#### 3. Inconsistencies with the NumPy API and Array API
+
+Although the Array API has managed to
+  get good coverage of the NumPy API, there are places where there are inconsistencies between
+  the two, it could be either with the function parameters or behaviour or the function or name
+  of the function like `concat` and `concatenate`.
+
+* Naming inconsistency between NumPy API and Array API
+
+```ipython
+
+In [1]: import numpy as np
+In [2]: import numpy.array_api as npx
+
+# Numpy API
+In [3]: np.concatenate
+Out[3]: <function numpy.concatenate>
+
+# Array API
+In [4]: npx.concat
+Out[4]: <function numpy.array_api._manipulation_functions.concat...>
+```
+
+* Indexing inconsistency between NumPy API and Array API
+
+```ipython
+In [1]: import numpy as np
+In [2]: import numpy.array_api as npx
+
+# Works with NumPy API
+In [3]: x = np.arange(10)
+In [4]: x[np.arange(2)]
+Out[4]: array([0, 1])
+
+
+# Doesn't Works with Array API
+In [5]: z = npx.arange(10)
+In [6]: z[npx.arange(2)]
+-----------------------------------------------------------------
+...
+IndexError: Non-zero dimensional integer array indices are not allowed in the array API namespace
+```
+
+## Results
+
+Here we will talk about about the performance of spectral clustering on GPU vs CPU
+
+We ran this on AMD as well as NVIDIA GPUs and here is a plot of the performance.
+The x-axis of the plot is the image dimension, i.e the dimension of the original
+image after rescaling it to a certain proportion, for this demo rescaled the image
+to 0.1, 0.2, 0.4, 0.6, 0.8 and 1x.
+
+
+#### On AMD GPU:
+
+![NumPy cs CuPy AMD](/images/2022/02/numpy_vs_cupy_amd.png)
+
+The plot for AMD GPU is not what you would expect, the computation is faster
+with numpy (i.e. on CPU) compared to cupy (i.e. on GPU) for all
+image sizes, this is due to the slow device synchronization issue on AMD GPUs.
```

Comment:
... which we are investigating and will submit a bug report to the CuPy project to discuss the strategies for fixing the problem.

Reply:
That's not worth tweaking I think - we need to fix or work around the issue before we can publish this blog post, right?

Reply:
We have not been able to isolate the synchronization issue based on my little understanding of AMD and cupy. I am happy to spend more time on this, if we're able to?

Reply:
@aktech let's discuss on Slack, I'll summarize there in a minute.